File: backend/src/services/authService.ts
import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';
import crypto from 'crypto';
import prisma from '../config/db';
import ErrorResponse from '../utils/errorResponse';
const getDeviceName = (userAgent: string | undefined): string => {
  if (!userAgent) return 'Unknown Device';
  let os = 'Unknown OS';
  let browser = 'Unknown Browser';
  if (userAgent.includes('Windows')) os = 'Windows';
  else if (userAgent.includes('iPhone')) os = 'iPhone';
  else if (userAgent.includes('iPad')) os = 'iPad';
  else if (userAgent.includes('Macintosh')) os = 'macOS';
  else if (userAgent.includes('Android')) os = 'Android';
  else if (userAgent.includes('Linux')) os = 'Linux';
  if (userAgent.includes('Firefox')) browser = 'Firefox';
  else if (userAgent.includes('Edge') || userAgent.includes('Edg')) browser = 'Edge';
  else if (userAgent.includes('Chrome')) browser = 'Chrome';
  else if (userAgent.includes('Safari')) browser = 'Safari';
  return `${os} (${browser})`;
};
const generateTokens = async (userId: string, userAgent: string | undefined, ipAddress: string | undefined) => {
  const rawRefreshToken = crypto.randomBytes(40).toString('hex');
  const expiresAt = new Date(Date.now() + 7 * 24 * 60 * 60 * 1000);
  const deviceName = getDeviceName(userAgent);
  const ip = ipAddress || 'Unknown IP';
  await prisma.session.deleteMany({
    where: {
      user_id: userId,
      expires_at: { lt: new Date() }
    }
  });
  const existingSession = await prisma.session.findFirst({
    where: {
      user_id: userId,
      device_name: deviceName,
      ip_address: ip
    }
  });
  let session;
  if (existingSession) {
    session = await prisma.session.update({
      where: { id: existingSession.id },
      data: {
        refresh_token: rawRefreshToken,
        expires_at: expiresAt,
        created_at: new Date()
      }
    });
  } else {
    session = await prisma.session.create({
      data: {
        user_id: userId,
        refresh_token: rawRefreshToken,
        device_name: deviceName,
        ip_address: ip,
        expires_at: expiresAt
      }
    });
  }
  const accessToken = jwt.sign({ id: userId, sessionId: session.id }, process.env.JWT_SECRET as string, {
    expiresIn: '15m'
  });
  return { accessToken, refreshToken: rawRefreshToken };
};
export const registerUser = async (name: string, email: string, password: string) => {
  const existingUser = await prisma.user.findUnique({
    where: { email }
  });
  if (existingUser) {
    throw new ErrorResponse('Email address is already registered.', 400);
  }
  const salt = await bcrypt.genSalt(10);
  const passwordHash = await bcrypt.hash(password, salt);
  const verificationToken = crypto.randomUUID();
  const newUser = await prisma.user.create({
    data: {
      name,
      email,
      password_hash: passwordHash,
      is_verified: false,
      verification_token: verificationToken
    },
    select: {
      id: true,
      name: true,
      email: true,
      created_at: true
    }
  });
  const verifyUrl = `http://localhost:5000/api/auth/verify?token=${verificationToken}`;
  console.log('\n==================================================');
  console.log('EMAIL VERIFICATION LINK');
  console.log(`Please click the link below to verify your account:`);
  console.log(`${verifyUrl}`);
  console.log('==================================================\n');
  return { user: newUser };
};
export const loginUser = async (email: string, password: string, userAgent: string | undefined, ipAddress: string | undefined) => {
  const user = await prisma.user.findUnique({
    where: { email }
  });
  if (!user) {
    throw new ErrorResponse('Invalid email or password.', 401);
  }
  const isMatch = await bcrypt.compare(password, user.password_hash);
  if (!isMatch) {
    throw new ErrorResponse('Invalid email or password.', 401);
  }
  if (!user.is_verified) {
    throw new ErrorResponse('Please verify your email before logging in. A verification link has been printed to the server console.', 401);
  }
  const { accessToken, refreshToken } = await generateTokens(user.id, userAgent, ipAddress);
  return {
    user: {
      id: user.id,
      name: user.name,
      email: user.email,
      created_at: user.created_at
    },
    accessToken,
    refreshToken
  };
};
export const refreshAccessToken = async (refreshToken: string) => {
  const session = await prisma.session.findUnique({
    where: { refresh_token: refreshToken }
  });
  if (!session || session.expires_at < new Date()) {
    if (session) {
      await prisma.session.delete({ where: { id: session.id } });
    }
    throw new ErrorResponse('Invalid or expired session. Please log in again.', 401);
  }
  const accessToken = jwt.sign({ id: session.user_id, sessionId: session.id }, process.env.JWT_SECRET as string, {
    expiresIn: '15m'
  });
  return { accessToken };
};
export const terminateSession = async (userId: string, sessionId: string) => {
  const session = await prisma.session.findUnique({
    where: { id: sessionId }
  });
  if (!session || session.user_id !== userId) {
    throw new ErrorResponse('Session not found.', 404);
  }
  await prisma.session.delete({
    where: { id: sessionId }
  });
};
export const terminateAllSessions = async (userId: string) => {
  await prisma.session.deleteMany({
    where: { user_id: userId }
  });
};
export const getActiveSessions = async (userId: string) => {
  return await prisma.session.findMany({
    where: { user_id: userId },
    orderBy: { created_at: 'desc' },
    select: {
      id: true,
      device_name: true,
      ip_address: true,
      created_at: true,
      refresh_token: true
    }
  });
};
export const forgotPassword = async (email: string) => {
  const user = await prisma.user.findUnique({ where: { email } });
  if (!user) {
    throw new ErrorResponse('No account found with that email.', 404);
  }
  const resetToken = crypto.randomBytes(32).toString('hex');
  const hashedToken = crypto.createHash('sha256').update(resetToken).digest('hex');
  const expires = new Date();
  expires.setHours(expires.getHours() + 1);
  await prisma.user.update({
    where: { email },
    data: {
      reset_token: hashedToken,
      reset_token_expires: expires
    }
  });
  const resetUrl = `http://localhost:5173/reset-password?token=${resetToken}`;
  console.log('\n==================================================');
  console.log('PASSWORD RESET LINK');
  console.log(`Requested for: ${email}`);
  console.log(`Please click the link below to reset your password:`);
  console.log(`${resetUrl}`);
  console.log('==================================================\n');
};
export const resetPassword = async (token: string, newPassword: string) => {
  const hashedToken = crypto.createHash('sha256').update(token).digest('hex');
  const user = await prisma.user.findFirst({
    where: {
      reset_token: hashedToken,
      reset_token_expires: { gt: new Date() }
    }
  });
  if (!user) {
    throw new ErrorResponse('Invalid or expired password reset token.', 400);
  }
  const salt = await bcrypt.genSalt(10);
  const passwordHash = await bcrypt.hash(newPassword, salt);
  await prisma.user.update({
    where: { id: user.id },
    data: {
      password_hash: passwordHash,
      reset_token: null,
      reset_token_expires: null
    }
  });
  await terminateAllSessions(user.id);
};
export const changePassword = async (userId: string, oldPassword: string, newPassword: string) => {
  const user = await prisma.user.findUnique({ where: { id: userId } });
  if (!user) throw new ErrorResponse('User not found.', 404);
  const isMatch = await bcrypt.compare(oldPassword, user.password_hash);
  if (!isMatch) {
    throw new ErrorResponse('Incorrect current password.', 401);
  }
  const salt = await bcrypt.genSalt(10);
  const passwordHash = await bcrypt.hash(newPassword, salt);
  await prisma.user.update({
    where: { id: userId },
    data: { password_hash: passwordHash }
  });
};

File: backend/src/main/java/com/urlshortener/service/AuthService.java
package com.urlshortener.service;

import com.urlshortener.entity.Session;
import com.urlshortener.entity.User;
import com.urlshortener.exception.ApiException;
import com.urlshortener.repository.SessionRepository;
import com.urlshortener.repository.UserRepository;
import com.urlshortener.util.JwtUtil;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class AuthService {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private SessionRepository sessionRepository;

    @Autowired
    private JwtUtil jwtUtil;

    @Autowired
    private PasswordEncoder passwordEncoder;

    private final SecureRandom secureRandom = new SecureRandom();

    @Transactional
    public Map<String, Object> registerUser(String name, String email, String password) {
        
        Optional<User> existing = userRepository.findByEmail(email);
        if (existing.isPresent()) {
            throw new ApiException("Email address is already registered.", 400);
        }

        String hashedPassword = passwordEncoder.encode(password);

        String verificationToken = UUID.randomUUID().toString();

        User user = new User();
        user.setName(name);
        user.setEmail(email);
        user.setPasswordHash(hashedPassword);
        user.setVerified(false);
        user.setVerificationToken(verificationToken);
        user.setCreatedAt(Instant.now());

        user = userRepository.save(user);

        System.out.println("Verification link: http://localhost:5000/api/auth/verify?token=" + verificationToken);

        Map<String, Object> userInfo = new LinkedHashMap<>();
        userInfo.put("id", user.getId());
        userInfo.put("name", user.getName());
        userInfo.put("email", user.getEmail());
        userInfo.put("created_at", user.getCreatedAt());

        return userInfo;
    }

    @Transactional
    public Map<String, Object> loginUser(String email, String password, String userAgent, String ipAddress) {
        
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new ApiException("Invalid email or password.", 401));

        if (!passwordEncoder.matches(password, user.getPasswordHash())) {
            throw new ApiException("Invalid email or password.", 401);
        }

        if (!user.isVerified()) {
            throw new ApiException("Please verify your email before logging in. Check your inbox for the verification link.", 401);
        }

        Map<String, String> tokens = generateTokens(user.getId(), userAgent, ipAddress);

        Map<String, Object> userInfo = new LinkedHashMap<>();
        userInfo.put("id", user.getId());
        userInfo.put("name", user.getName());
        userInfo.put("email", user.getEmail());
        userInfo.put("created_at", user.getCreatedAt());

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("user", userInfo);
        result.put("accessToken", tokens.get("accessToken"));
        result.put("refreshToken", tokens.get("refreshToken"));

        return result;
    }

    @Transactional
    private Map<String, String> generateTokens(UUID userId, String userAgent, String ipAddress) {
        
        byte[] randomBytes = new byte[40];
        secureRandom.nextBytes(randomBytes);
        String refreshToken = bytesToHex(randomBytes);

        Instant expiresAt = Instant.now().plus(7, ChronoUnit.DAYS);

        String deviceName = getDeviceName(userAgent);

        sessionRepository.deleteByUserIdAndExpiresAtBefore(userId, Instant.now());

        Optional<Session> existingSession = sessionRepository.findByUserIdAndDeviceNameAndIpAddress(userId, deviceName, ipAddress);

        Session session;
        if (existingSession.isPresent()) {
            
            session = existingSession.get();
            session.setRefreshToken(refreshToken);
            session.setExpiresAt(expiresAt);
            session.setCreatedAt(Instant.now());
            session = sessionRepository.save(session);
        } else {
            
            session = new Session();
            session.setUserId(userId);
            session.setRefreshToken(refreshToken);
            session.setDeviceName(deviceName);
            session.setIpAddress(ipAddress);
            session.setExpiresAt(expiresAt);
            session.setCreatedAt(Instant.now());
            session = sessionRepository.save(session);
        }

        String accessToken = jwtUtil.generateAccessToken(userId, session.getId());

        Map<String, String> tokens = new LinkedHashMap<>();
        tokens.put("accessToken", accessToken);
        tokens.put("refreshToken", refreshToken);

        return tokens;
    }

    private String getDeviceName(String userAgent) {
        if (userAgent == null || userAgent.isEmpty()) {
            return "Unknown OS (Unknown Browser)";
        }

        String os;
        if (userAgent.contains("Windows")) {
            os = "Windows";
        } else if (userAgent.contains("iPhone")) {
            os = "iPhone";
        } else if (userAgent.contains("iPad")) {
            os = "iPad";
        } else if (userAgent.contains("Macintosh") || userAgent.contains("Mac OS")) {
            os = "macOS";
        } else if (userAgent.contains("Android")) {
            os = "Android";
        } else if (userAgent.contains("Linux")) {
            os = "Linux";
        } else {
            os = "Unknown OS";
        }

        String browser;
        if (userAgent.contains("Firefox")) {
            browser = "Firefox";
        } else if (userAgent.contains("Edg")) {
            browser = "Edge";
        } else if (userAgent.contains("Chrome") || userAgent.contains("Chromium")) {
            browser = "Chrome";
        } else if (userAgent.contains("Safari")) {
            browser = "Safari";
        } else {
            browser = "Unknown Browser";
        }

        return os + " (" + browser + ")";
    }

    @Transactional
    public Map<String, Object> refreshAccessToken(String refreshToken) {
        
        Session session = sessionRepository.findByRefreshToken(refreshToken)
                .orElseThrow(() -> new ApiException("Invalid or expired session. Please log in again.", 401));

        if (session.getExpiresAt().isBefore(Instant.now())) {
            sessionRepository.delete(session);
            throw new ApiException("Invalid or expired session. Please log in again.", 401);
        }

        String accessToken = jwtUtil.generateAccessToken(session.getUserId(), session.getId());

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("accessToken", accessToken);

        return result;
    }

    public List<Map<String, Object>> getActiveSessions(UUID userId) {
        List<Session> sessions = sessionRepository.findByUserIdOrderByCreatedAtDesc(userId);

        return sessions.stream().map(session -> {
            Map<String, Object> sessionMap = new LinkedHashMap<>();
            sessionMap.put("id", session.getId());
            sessionMap.put("device_name", session.getDeviceName());
            sessionMap.put("ip_address", session.getIpAddress());
            sessionMap.put("created_at", session.getCreatedAt());
            sessionMap.put("refresh_token", session.getRefreshToken());
            return sessionMap;
        }).collect(Collectors.toList());
    }

    @Transactional
    public void terminateSession(UUID userId, String sessionId) {
        UUID sessionUUID;
        try {
            sessionUUID = UUID.fromString(sessionId);
        } catch (IllegalArgumentException e) {
            throw new ApiException("Session not found.", 404);
        }

        Session session = sessionRepository.findById(sessionUUID)
                .orElseThrow(() -> new ApiException("Session not found.", 404));

        if (!session.getUserId().equals(userId)) {
            throw new ApiException("Session not found.", 404);
        }

        sessionRepository.delete(session);
    }

    @Transactional
    public void terminateAllSessions(UUID userId) {
        sessionRepository.deleteByUserId(userId);
    }

    @Transactional
    public void forgotPassword(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new ApiException("No account found with that email.", 404));

        byte[] randomBytes = new byte[32];
        secureRandom.nextBytes(randomBytes);
        String rawToken = bytesToHex(randomBytes);

        String hashedToken = sha256Hash(rawToken);

        user.setResetToken(hashedToken);
        user.setResetTokenExpires(Instant.now().plus(1, ChronoUnit.HOURS));
        userRepository.save(user);

        System.out.println("Password reset link: http://localhost:5173/reset-password?token=" + rawToken);
    }

    @Transactional
    public void resetPassword(String token, String newPassword) {
        
        String hashedToken = sha256Hash(token);

        User user = userRepository.findByResetToken(hashedToken)
                .orElseThrow(() -> new ApiException("Invalid or expired password reset token.", 400));

        if (user.getResetTokenExpires() == null || user.getResetTokenExpires().isBefore(Instant.now())) {
            throw new ApiException("Invalid or expired password reset token.", 400);
        }

        user.setPasswordHash(passwordEncoder.encode(newPassword));
        user.setResetToken(null);
        user.setResetTokenExpires(null);
        userRepository.save(user);

        terminateAllSessions(user.getId());
    }

    @Transactional
    public void changePassword(UUID userId, String oldPassword, String newPassword) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ApiException("User not found.", 404));

        if (!passwordEncoder.matches(oldPassword, user.getPasswordHash())) {
            throw new ApiException("Incorrect current password.", 401);
        }

        user.setPasswordHash(passwordEncoder.encode(newPassword));
        userRepository.save(user);
    }

    private String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }

    private String sha256Hash(String input) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(input.getBytes(StandardCharsets.UTF_8));
            return bytesToHex(hash);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-256 algorithm not available", e);
        }
    }
}
